### PR TITLE
nixos/pam: Type all limit options

### PR DESCRIPTION
**Pull request by dasJ**
###### Motivation for this change

#76184

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/security/pam.nix [hunk 1]
```diff
@@ -648,6 +650,51 @@ let
...
+
+      type = mkOption {
+        description = "Type of this limit";
+        type = enum [ "-" "hard" "soft" ];
+        default = "-";
+      };
+
+      item = mkOption {
+        description = "Item this limit applies to";
+        type = enum [
```
Comment (flokli) [inline nixos/modules/security/pam.nix:669]: Are we sure this list won't need to be extended?

File: nixos/modules/security/pam.nix [hunk 2]
```diff
@@ -295,6 +295,8 @@ let
       };
 
       limits = mkOption {
+        default = [];
+        type = limitsType;
         description = ''
           Attribute set describing resource limits.  Defaults to the
           value of <option>security.pam.loginLimits</option>.
```
Comment (flokli) [inline nixos/modules/security/pam.nix:302]: We might want to have a link to some docs here, so the choices below make a bit more sense.

File: nixos/modules/security/pam.nix [hunk 3]
```diff
@@ -648,6 +650,51 @@ let
...
+    options = {
+      domain = mkOption {
+        description = "Username, groupname, or wildcard this limit applies to";
+        example = "@wheel";
+        type = str;
+      };
+
+      type = mkOption {
+        description = "Type of this limit";
+        type = enum [ "-" "hard" "soft" ];
```
Comment (flokli) [inline nixos/modules/security/pam.nix:663]: What's `-` here? Might need another comment - or maybe some links to docs.

**Review by flokli — COMMENTED** [review]
This needs a bit more docs, or pointers to upstream documentation. I don't really understand much from just reading the descriptions.

Also, I'm a bit worried about all the `enums` - are we sure that list is static enough?

File: nixos/modules/security/pam.nix (hunk 1 above)
Comment (dasJ) [reply nixos/modules/security/pam.nix:669]: Pretty sure the list of ulimits won't be extended anytime soon

File: nixos/modules/security/pam.nix (hunk 3 above)
Comment (dasJ) [reply nixos/modules/security/pam.nix:666]: I linked to the docs in the main option

**Pull request merged**